### ISSUE TITLE
A lesson about DDP and computation graph

### ISSUE THREAD

**Issue by harryhan618** [harryhan618/LaneNet#12]
This is a lesseon I learnt about `DistributedDataParallel` and forward & backward computation graph. So I would like record it here.
The lesson is: when calculating the loss, always use 0/1 mask to multiply the loss result for valid pixel select, rather than use bool mask for valid pixel indexing. Even though this would take more memory during training (I guess).

Here's the case. When I want to calculate a loss that I implement myself, say `discriminative_loss` in this repo, I would ignore it if there is no positive sample. That is to say:
```
loss_all = 0
if len(torch.unique(seg_gt))==0:  # only background pixel, so I don't calculate loss
  continue
else:
  loss_some_task = calculate_loss(pred, seg_gt)
  loss_all = loss_all + loss_some_task
```
Thus in this case, when a sample has no positive background, `pred` won't get involve into computation graph for calculating loss.
Ok, here is the tricky part. When using `DDP`, it expects the output of the model should either involved in the computation graph or not across all processes. So a special case happens when `pred` is involved in the loss compuation graph in Process#1 and otherwise in Process#2. This will make both Processes stuck.

So always let the output get involved in to loss compuation graph. I would revise the code above like this:
```
loss_all = 0
if len(torch.unique(seg_gt))==0:  # only background pixel, so I don't calculate loss
  _nonsense_sum = pred.sum()
  _nonsense_zero = torch.zeros_like(_nonsense_sum)
  loss_all = loss_all + _nonsense_sum * _nonsense_zero
else:
  loss_some_task = calculate_loss(pred, seg_gt)
  loss_all = loss_all + loss_some_task
```

**Comment by ginobilinie** [comment]
Very helpful. I also encountered such a problem in other scenarios. According to my experience, apex.DDP can avoid such a problem. DDP provided by pytorch will have such a issue.
